crates/client/state-sync/src/ethereum.rs

```diff
@@ -0,0 +1,668 @@
+
+        Ok(Self {
+            http_provider: provider,
+            core_contract,
+            verifier_contract,
+            memory_page_contract,
+            eth_url_list: eth_urls,
+            current_provider_index: Arc::new(Mutex::new(0)),
+            sync_status,
+            v011_diff_format_height,
+            constructor_args_diff_height,
+        })
+    }
+}
+
+#[cfg(test)]
+impl EthereumStateFetcher<MockProvider> {
+    pub fn mock(
+        core_contract: Address,
+        verifier_contract: Address,
+        memory_page_contract: Address,
+        eth_url_list: Vec<String>,
+        v011_diff_format_height: u64,
+        mock_provider: MockProvider,
+        sync_status: Arc<Mutex<SyncStatus>>,
+        constructor_args_diff_height: u64,
+    ) -> Result<Self, Error> {
+        let provider = Provider::<MockProvider>::new(mock_provider);
+
+        Ok(Self {
+            http_provider: provider,
+            core_contract,
+            verifier_contract,
+            memory_page_contract,
+            eth_url_list,
+            current_provider_index: Arc::new(Mutex::new(0)),
+            sync_status,
+            v011_diff_format_height,
+            constructor_args_diff_height,
+        })
+    }
+}
+
+impl<P: JsonRpcClient + Clone> EthereumStateFetcher<P> {
+    pub async fn get_logs_retry(&mut self, filter: &Filter) -> Result<Vec<Log>, Error> {
+        if let Ok(res) = self.http_provider.get_logs(filter).await {
+            return Ok(res);
+        }
+
+        let mut retries = 0;
+        loop {
+            let provider = self
+                .current_provider_index
+                .lock()
+                .map(|index| {
+                    let current_provider_url = &self.eth_url_list[*index];
+                    Provider::<Http>::try_from(current_provider_url).map_err(|e| Error::L1Connection(e.to_string()))
+                })
+                .map_err(|e| Error::Other(e.to_string()))??;
+
+            match provider.get_logs(filter).await {
+                Ok(logs) => return Ok(logs),
+                Err(_e) => {
+                    retries += 1;
+                    if retries > self.eth_url_list.len() {
+                        return Err(Error::L1Connection("All Ethereum nodes failed.".to_string()));
+                    }
+
+                    // change to next Ethereum node
+                    if let Ok(mut index) = self.current_provider_index.lock() {
+                        *index = (*index + 1) % self.eth_url_list.len();
+                    };
+
+                    // Calculate the wait time manually
+                    let wait_time = self.calculate_backoff(retries);
+
+                    sleep(wait_time).await;
+                }
+            }
+        }
+    }
+
+    // Custom backoff calculation
+    pub fn calculate_backoff(&self, retries: usize) -> Duration {
+        // A simple exponential backoff with a maximum delay of 10 seconds
```

Comment:
The calculate_backoff method employs a simple exponential backoff strategy, where the waiting time increases exponentially with the number of retries. Here is a detailed description of the method:

- base_delay is the initial waiting time for each retry, set to 1 second.
- exponential_factor is the exponential factor, set to 2.0, indicating that the waiting time will double with each retry.
- max_delay is the maximum allowed waiting time, set to 10 seconds, to prevent an unlimited growth of waiting time.

Certainly, the max_delay value is an estimated time based on user experience. Perhaps you have more reasonable suggested values.